### PR TITLE
Discord rich presence integration

### LINKED ISSUE #3547: [Feature Request] Add Discord "Rich Presence" Integration
Citra is already a verified app with Discord, so why not take it further? I request integration with "Rich Presence" so other people on Discord can see what 3DS game you're playing, and for how long. Here's some docs on integrating it: https://discordapp.com/developers/docs/rich-presence/how-to

### PR DESCRIPTION
**Pull request by valentinvanelslande**
Screenshot:
screenshot
Closes #3547

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/3795)

### CODE REVIEW

**Comment by jroweboy** [comment]
Please don't link discord against core. Its way more acceptable to link it into the frontend where we can do everything you did here in addition to driving multiplayer lobbies and joining.

File: src/citra/citra.cpp [hunk 1]
```diff
@@ -323,9 +324,15 @@ int main(int argc, char** argv) {
         }
     }
 
+    std::string title;
+    system.GetAppLoader().ReadTitle(title);
+    DiscordRPC::Init();
```
Comment (j-selby) [inline src/citra/citra.cpp]: This should be done via an "integrations" abstraction - when the next Discord rolls around, we should have a proper framework in place.

Comment (j-selby) [reply src/citra/citra.cpp]: Further, I don't think this logic should be duplicated with that in `citra_qt`

File: src/core/frontend/discord.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+s64 g_start_time;
+bool g_send_presence = true;
+
+static void HandleDiscordReady(const DiscordUser* request) {}
+
+static void HandleDiscordDisconnected(int errorCode, const char* message) {}
+
+static void HandleDiscordError(int errorCode, const char* message) {}
+
+static void HandleDiscordJoin(const char* joinSecret) {}
```
Comment (j-selby) [inline src/core/frontend/discord.cpp]: Joining should be implemented, else this is just fluff.

File: src/core/frontend/discord.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    g_start_time = time(NULL);
+    g_send_presence = true;
+    DiscordEventHandlers handlers{};
+    handlers.ready = HandleDiscordReady;
+    handlers.disconnected = HandleDiscordDisconnected;
+    handlers.errored = HandleDiscordError;
+    handlers.joinGame = HandleDiscordJoin;
+    handlers.spectateGame = HandleDiscordSpectate;
+    handlers.joinRequest = HandleDiscordJoinRequest;
+    Discord_Initialize("432553040755556374", &handlers, 1, NULL);
```
Comment (j-selby) [inline src/core/frontend/discord.cpp]: Don't reuse my test Discord application ID - seriously? Please have a talk with Flame Sage about putting this under CitraBot's account.

File: src/core/frontend/discord.cpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#include "core/frontend/discord.h"
+
+namespace DiscordRPC {
+
+s64 g_start_time;
+bool g_send_presence = true;
+
+static void HandleDiscordReady(const DiscordUser* request) {}
+
+static void HandleDiscordDisconnected(int errorCode, const char* message) {}
```
Comment (j-selby) [inline src/core/frontend/discord.cpp]: Error messages seem important to log, no?

File: src/core/frontend/discord.cpp [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+static void HandleDiscordError(int errorCode, const char* message) {}
```
Comment (j-selby) [inline src/core/frontend/discord.cpp]: Ditto.

File: src/core/frontend/discord.cpp (hunk 3 above)
Comment (chris062689) [reply src/core/frontend/discord.cpp]: I would feel much more comfortable for him to generate his own for testing until we're ready to push this to Canary.

**Comment by j-selby** [comment]
@wwylele Off by default? Particular reason for this?

**Comment by wwylele** [comment]
@j-selby Because I don't want to build it when I develop for citra (for less build time and so that others won't see me "playing" citra) and I don't want to add additional cmake configuration for that (IDE can be buggy and reset cmake configuration during update). We can still add the cmake flag to enable it for release build in CI.

**Comment by valentinvanelslande** [comment]
The new version of the PR (with comments addressed, not pushed) conflicts with #3776, closing.

**Pull request closed without merging**